usb: device_next: changes around descriptor handling, preparation for the USB device class skeleton

jfischer-no opened the pull request:
Changes around descriptor handling, preparation for the USB device class skeleton

Commit 15b8df1edebf07020ee2e2b2c672267c10e7daca fixes: #58168

jfischer-no committed:
usb: device_next: remove foobaz interface from common shell code

Foobaz interface will be renamed and added as a standalone part
to serve as an API example in a subsequent commit.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: do a little more cleanup on shutdown

At the shutdown of USB device stack we have to cleanup and
remove all registered class instances and string descriptors
from a configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow reuse of string descriptor nodes

If we try to reuse a string descriptor node, the content will
be corrupted because the device stack assumes it is still ASCII7
encoded. Add a flag to indicate that a descriptor node contains
UTF16LE encoded content.

And while we are at it, add a flag to indicate that the SN string
should not be obtained from the hwinfo API.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow selection of alternative USB device context

When we enable USBD shell support with samples like cdc_acm or mass,
there would be another USB device context besides the one provided
by the shell. This patch introduces a new command to select an
alternate context to be used by USBD shell commands.

Also fix the type where "add" command should be "remove".

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: use dlist instead of slist for descriptors list

While this does not provide much of an advantage yet,
it will allow us to add descriptors and assign an index
more easily in the next commit.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: use specific macros for string descriptors

Add and use specific macros for manufacturer, product, and
serial number string descriptors.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: allow string descriptor index to be updated

Support for multiple instances of a class implementation,
and the ability to register an instance to a configuration
at runtime, requires a mechanism to add a string descriptor
and update its index based on the total number of descriptors.

We also need to handle some special string descriptors like
Product or Serial Number provided by the application.
Marked as such by using specific macros, these descriptors
can be sorted out by the stack and the device descriptor
indexes are updated automatically.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: device_next: implement usbd_class_shutdown()

Implement marked as TODO usbd_class_shutdown().

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on include/zephyr/usb/usbd.h:
```diff
@@ -62,7 +62,7 @@ extern "C" {
 struct usbd_desc_node {
 	/** slist node struct */
-	sys_snode_t node;
+	sys_dnode_t node;
```
Where are you using the features that dlist has that are not found in slist? The commit message mentions "next commit" but it is not in next commit and even in the last commit I didn't find (maybe I just missed) where the dlist is really needed over slist.

tmon-nordic commented on subsys/usb/device_next/usbd_core.c:
```diff
@@ -199,6 +201,23 @@ int usbd_device_init_core(struct usbd_contex *const uds_ctx)
+	SYS_SLIST_FOR_EACH_CONTAINER(&uds_ctx->configs, cfg_nd, node) {
+		uint8_t cfg_value = usbd_config_get_value(cfg_nd);
+
+		ret = usbd_class_remove_all(uds_ctx, cfg_value);
```
I think the commits are overly separated. I would prefer if the helpers were added and used in the same commit.

jfischer-no replied:
They actually are, https://github.com/zephyrproject-rtos/zephyr/pull/58187/commits/6e1fdff22981fdcbcdec6fe91d57df7e1b14f59d. Do you think I should squash it with the last commit?

tmon-nordic replied:
I think one commit for the helpers + use helpers, keep the last commit separate.

jfischer-no committed:
usb: device_next: check if another context is initialized

There may be more than one context using the same device, for example,
if the shell module is enabled. Check if another context that uses
the same device is initialized.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
